### ISSUE TITLE
`toJSON` not respected in honeybadger sanitize

### ISSUE THREAD

**Issue by thatsmydoing** [honeybadger-io/honeybadger-js#722]
## What are the steps to reproduce this issue?
1. Add breadcrumb with metadata containing an object with a `toJSON`
2. Send a honeybadger notice

## What happens?
The metadata field in the breadcrumb just says `[object Object]`

## What were you expecting to happen?
It would use `toJSON` to expand it and then sanitize

## Any other comments?
The reason we want to do this is we do our own sanitization and eventual stringify of the breadcrumb metadata field but we don't want to do this eagerly since it wastes time and only want to do it at the point of reporting the error.

## What versions are you using?
**Operating System:** Linux
**Package Version:** 3.2.1
**Browser Version:** Firefox 98

**Comment by subzero10** [comment]
Hey @thatsmydoing, thanks for submitting an issue. We will look at this asap.

**Comment by subzero10** [comment]
Hey @thatsmydoing, I looked into this and it works as expected for me. Here's a report with a json object in the metadata field:
```
app.get('/notify', (req, res) => {
  Honeybadger.addBreadcrumb('breadcrumb with metadata', {
    metadata: {
      user: 'John Doe',
      aNumber: 345,
      aBoolean: true,
      innerObject: {
        anotherUser: 'John Smith',
      },
    }
  })
  Honeybadger.notify('report with breadcrumb')
  res.send('Notification sent!')
})
```
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/5813382/160912307-86d0f10b-3187-4836-b9dd-c5e87089ff61.png">

This points me to believe that the issue may be with the `toJSON` method you mentioned above. Can you please some of your code? Or even a repository where you are having this issue?

On a similar note, I've noticed that inner objects are represented as strings and this is something that we could improve.
What do you think @joshuap ?

**Comment by joshuap** [comment]
> Hey @thatsmydoing, I looked into this and it works as expected for me. Here's a report with a json object in the metadata field:
> 
> ```
> app.get('/notify', (req, res) => {
>   Honeybadger.addBreadcrumb('breadcrumb with metadata', {
>     metadata: {
>       user: 'John Doe',
>       aNumber: 345,
>       aBoolean: true,
>       innerObject: {
>         anotherUser: 'John Smith',
>       },
>     }
>   })
>   Honeybadger.notify('report with breadcrumb')
>   res.send('Notification sent!')
> })
> ```
> 
> <img alt="image" width="1142" src="https://user-images.githubusercontent.com/5813382/160912307-86d0f10b-3187-4836-b9dd-c5e87089ff61.png">
> 
> This points me to believe that the issue may be with the `toJSON` method you mentioned above. Can you please some of your code? Or even a repository where you are having this issue?
> 
> On a similar note, I've noticed that inner objects are represented as strings and this is something that we could improve. What do you think @joshuap ?

I know we intentionally decided to limit the nesting level of metadata for breadcrumbs, I think because context is a better use case if you really need to inspect nested data structures. @rabidpraxis, do you recall other reasons?

**Comment by thatsmydoing** [comment]
Sorry, I was unclear. I meant something like

```
  Honeybadger.addBreadcrumb('breadcrumb with metadata', {
    metadata: {
      foo: {
        toJSON() { return 'bar'; }
      }
    }
  })
  Honeybadger.notify('report with breadcrumb')
```

`JSON.stringify` by itself will just have foo be `bar` but the sanitize step of honeybadger turns it into `{"toJSON"=>"[object Function]"}`

> On a similar note, I've noticed that inner objects are represented as strings and this is something that we could improve.

I had asked about this and the response was that `JSON.stringify` with a space is how you'd get this to show up nicely. This is partly a reason why we want `toJSON` supported so we don't have to stringify eagerly.

**Comment by subzero10** [comment]
@thatsmydoing Now I see, I will check this out and come back to you.

**Comment by subzero10** [comment]
@thatsmydoing This has been fixed! You can try it out with `v3.2.9` or `v4.0.0-beta.2`.

**Comment by thatsmydoing** [comment]
Thanks! This works for us. I notice that the implementation simply lets the `toJSON` function through which means the depth limit wouldn't be respected for anything returned by `toJSON`. We only return strings so this doesn't matter for us but it could be a gotcha.
